### PR TITLE
Introduce fallable handlers

### PR DESCRIPTION
**Pull request by steveklabnik**
If your request fails, we return a 500 internal server error.

This is useful so that things built on top can propogate errors if
they don't care about handling them specifically, making the server
overall more robust.

This should help with https://github.com/steveklabnik/simple-server/issues/21

### CODE REVIEW

File: examples/routes.rs [hunk 1]
```diff
@@ -1,28 +1,24 @@
...
-                response_builder.body("<h1>404</h1><p>Not found!<p>".as_bytes()).unwrap()
+                response.status(StatusCode::NOT_FOUND);
+                Ok(response.body("<h1>404</h1><p>Not found!<p>".as_bytes())?)
             }
         }
-    };
+    });
 
-    let server = Server::new(handler);
+    let host = "127.0.0.1";
```
Comment (ashleygwilliams) [inline examples/routes.rs]: prefer to define these at the top, we should also make it take an optional env var. can do in another PR

File: examples/simple-server.rs [hunk 2]
```diff
@@ -5,7 +5,7 @@ use simple_server::Server;
 fn main() {
     let server = Server::new(|request, mut response| {
         println!("Request received. {} {}", request.method(), request.uri());
-        response.body("Hello Rust!".as_bytes()).unwrap()
+        Ok(response.body("Hello Rust!".as_bytes())?)
```
Comment (ashleygwilliams) [inline examples/simple-server.rs]: do we love this syntax?

File: src/lib.rs [hunk 3]
```diff
@@ -68,7 +69,13 @@ impl Server {
             return;
         }
 
-        let response = (self.handler)(request, response_builder);
+        let response = (self.handler)(request, response_builder).unwrap_or_else(|_| {
```
Comment (ashleygwilliams) [inline src/lib.rs]: i'd like this error in a separate file if we could.

Comment (ashleygwilliams) [reply src/lib.rs]: maybe a `status` helper?

**Comment by ashleygwilliams** [comment]
this also needs a test that proves the 500

File: examples/routes.rs (hunk 1 above)
Comment (steveklabnik) [reply examples/routes.rs]: yeah i guess this is one preference against another; I don't care strongly enough though, so we can go with yours

File: examples/simple-server.rs (hunk 2 above)
Comment (steveklabnik) [reply examples/simple-server.rs]: Are you referring to `Ok`, the `?`, or `as_bytes`

File: src/lib.rs (hunk 3 above)
Comment (steveklabnik) [reply src/lib.rs]: I'm not sure what this means.

File: examples/simple-server.rs (hunk 2 above)
Comment (ashleygwilliams) [reply examples/simple-server.rs]: the `Ok`, `?`

**Comment by ashleygwilliams** [comment]
rally jsut waiting on the test, otherwise we can merge this and deal with the other comments

File: src/lib.rs (hunk 3 above)
Comment (ashleygwilliams) [reply src/lib.rs]: i want to get the html out of this, basically. i can write something up later.

**Comment by ashleygwilliams** [comment]
i'd like to see the test in the same file as our current test. with a name that explains what it does. i think your file name was a good one.
